ceno_zkvm/src/instructions/riscv/div.rs

```diff
@@ -1,124 +1,348 @@
-        let mut divisor = UInt::new_unchecked(|| "divisor", cb)?;
-        let mut outcome = UInt::new(|| "outcome", cb)?;
-        let r = UInt::new(|| "remainder", cb)?;
-        let (dividend, inter_mul_value) =
-            divisor.mul_add(|| "divisor * outcome + r", cb, &mut outcome, &r, true)?;
-
-        // div by zero check
-        let is_zero =
-            IsZeroConfig::construct_circuit(cb, || "divisor_zero_check", divisor.value())?;
-        let outcome_value = outcome.value();
+        // The soundness analysis for these constraints is only valid for
+        // 32-bit registers represented over the Goldilocks field, so verify
+        // these parameters
+        assert_eq!(UInt::<E>::TOTAL_BITS, u32::BITS as usize);
+        assert_eq!(E::BaseField::MODULUS_U64, goldilocks::MODULUS);
+
+        // 32-bit value from rs1
+        let dividend = UInt::new_unchecked(|| "dividend", cb)?;
+        // 32-bit value from rs2
+        let divisor = UInt::new_unchecked(|| "divisor", cb)?;
+        let quotient = UInt::new(|| "quotient", cb)?;
+        let remainder = UInt::new(|| "remainder", cb)?;
+
+        // `rem_e` and `div_e` are expressions representing the remainder and
+        // divisor from the signed or unsigned division operation, with signs
+        // normalized to be nonnegative, so that correct values must satisfy
+        // either `0 <= rem_e < div_e` or `div_e == 0`.  The `rem_e` value
+        // should be constrained to be nonnegative before being returned from
+        // this block, while the checks `rem_e < div_e` or `div_e == 0` are
+        // done later.
+        let (internal_config, rem_e, div_e) = match I::INST_KIND {
+            InsnKind::DIVU | InsnKind::REMU => {
+                cb.require_equal(
+                    || "unsigned_division_relation",
+                    dividend.value(),
+                    divisor.value() * quotient.value() + remainder.value(),
+                )?;
+
+                (InternalDivRem::Unsigned, remainder.value(), divisor.value())
+            }
+
+            InsnKind::DIV | InsnKind::REM => {
+                let dividend_signed =
+                    Signed::construct_circuit(cb, || "dividend_signed", &dividend)?;
+                let divisor_signed = Signed::construct_circuit(cb, || "divisor_signed", &divisor)?;
+                let quotient_signed =
+                    Signed::construct_circuit(cb, || "quotient_signed", &quotient)?;
+                let remainder_signed =
+                    Signed::construct_circuit(cb, || "remainder_signed", &remainder)?;
+
+                // The quotient and remainder can be interpreted as non-positive
+                // values when exactly one of dividend and divisor is negative
+                let neg_div_expr = {
+                    let a_neg = dividend_signed.is_negative.expr();
+                    let b_neg = divisor_signed.is_negative.expr();
+                    &a_neg * (1 - &b_neg) + (1 - &a_neg) * &b_neg
```

Comment:
Really minor one, I like the original style more because It's clear and slightly efficient. It reduces the number of recursion [here](https://github.com/scroll-tech/ceno/blob/master/ceno_zkvm/src/expression/monomial.rs#L13). But it depends on you :)
```rust                    
// a_neg * (1 - b_neg) + (1 - a_neg) * b_neg
(a_neg.clone() + b_neg.clone()) - (Expression::<E>::from(2) * a_neg * b_neg)
```

Reply:
Yeah, I think the fact that it was simplified to monomial form was why I went with the original version, but then the current revision is closer to the underlying logic of the expression, which gives an easily readable meaning for binary inputs.  I don't think I have a strong opinion about it -- any thoughts @matthiasgoergens since the change was your suggestion?

Reply:
There are two parts to my suggestion.  One is to remove the manual clones, and the other is to use a simpler algebraic expression.

They are independent of each other.

I don't really have much of an opinion on whether `(a + b - 2 * a * b)` or `a * (1 - b) + (1 - a) * b` is easier to read.  I went with the latter, because someone had made a comment `// a_neg * (1 - b_neg) + (1 - a_neg) * b_neg` to that end; so I assume someone thought that form is easier to read.

Here's my suggestion, if you want to use the original expression:

```rust
// a_neg ^ b_neg
&a_neg + &b_neg - 2 * a_neg * b_neg
```

Btw, when you say 'efficient', do you mean that the distribution function will be called slightly less often; or do you mean that the final resulting expression will be different?

Reply:
Sure, this seems also like a good way to write it, with the simple comment -- updated to this version in the latest commit.